File: iOS/TRTCScenesDemo/TXLiteAVDemo/TRTCVoiceRoomDemo/model/Impl/trtc/VoiceRoomTRTCService.m
//
//  VoiceRoomTRTCService.m
//  TRTCVoiceRoomOCDemo
//
//  Created by abyyxwang on 2020/7/1.
//  Copyright © 2020 tencent. All rights reserved.
//

#import "VoiceRoomTRTCService.h"
#import "TRTCCloud.h"

@interface VoiceRoomTRTCService () <TRTCCloudDelegate>

@property (nonatomic, assign) BOOL isInRoom;
@property (nonatomic, strong) NSString *userId;
@property (nonatomic, strong) NSString *roomId;
@property (nonatomic, strong) TRTCParams *mTRTCParms;
@property (nonatomic, copy) TXCallback enterRoomCallback;
@property (nonatomic, copy) TXCallback exitRoomCallback;

@property (nonatomic, strong, readonly)TRTCCloud *mTRTCCloud;

@end

@implementation VoiceRoomTRTCService

- (TRTCCloud *)mTRTCCloud {
    return [TRTCCloud sharedInstance];
}

+ (instancetype)sharedInstance{
    static VoiceRoomTRTCService* instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[VoiceRoomTRTCService alloc] init];
    });
    return instance;
}

- (void)enterRoomWithSdkAppId:(UInt32)sdkAppId
                       roomId:(NSString *)roomId
                       userId:(NSString *)userId
                     userSign:(NSString *)userSign
                         role:(NSInteger)role
                     callback:(TXCallback)callback {
    BOOL isParamError = NO;
    if (roomId == nil || [roomId isEqualToString:@""]) {
        isParamError = YES;
    }
    if (userId == nil || [userId isEqualToString:@""]) {
        isParamError = YES;
    }
    if (userSign == nil || [userSign isEqualToString:@""]) {
        isParamError = YES;
    }
    int roomIdIntValue = [roomId intValue];
    if (roomIdIntValue == 0) {
        isParamError = YES;
    }
    if (isParamError) {
        TRTCLog(@"error: enter trtc room fail. params invalid. room id:%@, userId:%@, userSig is empty:%d", roomId, userId, (userSign == nil || [userSign isEqualToString:@""]));
        callback(-1, @"enter trtc room fail.");
        return;
    }
    self.userId = userId;
    self.roomId = roomId;
    self.enterRoomCallback = callback;
    TRTCLog(@"enter room. app id:%u, room id: %@, userID: %@", (unsigned int)sdkAppId, roomId, userId);
    TRTCParams * parms = [[TRTCParams alloc] init];
    parms.sdkAppId = sdkAppId;
    parms.userId = userId;
    parms.userSig = userSign;
    parms.role = role == 20 ? TRTCRoleAnchor : TRTCRoleAudience;
    parms.roomId = roomIdIntValue;
    self.mTRTCParms = parms;
    [self internalEnterRoom];
}

- (void)exitRoom:(TXCallback)callback {
    TRTCLog(@"exit trtc room.");
    self.userId = nil;
    self.mTRTCParms = nil;
    self.enterRoomCallback = nil;
    self.exitRoomCallback = callback;
    [self.mTRTCCloud exitRoom];
}

- (void)muteLocalAudio:(BOOL)isMute {
    [self.mTRTCCloud muteLocalAudio:isMute];
}

- (void)muteRemoteAudioWithUserId:(NSString *)userId isMute:(BOOL)isMute {
    [self.mTRTCCloud muteRemoteAudio:userId mute:isMute];
}

- (void)muteAllRemoteAudio:(BOOL)isMute {
    [self.mTRTCCloud muteAllRemoteAudio:isMute];
}

- (void)setAudioQuality:(NSInteger)quality {
    TRTCAudioQuality targetQuality = TRTCAudioQualityDefault;
    switch (quality) {
        case 1:
            targetQuality = TRTCAudioQualitySpeech;
            break;
        case 3:
            targetQuality = TRTCAudioQualityMusic;
        default:
            break;
    }
    [self.mTRTCCloud setAudioQuality:targetQuality];
}

- (void)startMicrophone {
    [self.mTRTCCloud startLocalAudio];
}

- (void)stopMicrophone {
    [self.mTRTCCloud stopLocalAudio];
}

- (void)switchToAnchor {
    [self.mTRTCCloud switchRole:TRTCRoleAnchor];
    [self.mTRTCCloud startLocalAudio];
}

- (void)switchToAudience {
    [self.mTRTCCloud stopLocalAudio];
    [self.mTRTCCloud switchRole:TRTCRoleAudience];
}

- (void)setSpeaker:(BOOL)userSpeaker {
    [self.mTRTCCloud setAudioRoute:userSpeaker ? TRTCAudioModeSpeakerphone : TRTCAudioModeEarpiece];
}

- (void)setAudioCaptureVolume:(NSInteger)volume {
    [self.mTRTCCloud setAudioCaptureVolume:volume];
}

- (void)setAudioPlayoutVolume:(NSInteger)volume {
    [self.mTRTCCloud setAudioPlayoutVolume:volume];
}

- (void)startFileDumping:(TRTCAudioRecordingParams *)params {
    [self.mTRTCCloud startAudioRecording:params];
}

- (void)stopFileDumping {
    [self.mTRTCCloud stopAudioRecording];
}

- (void)enableAudioEvalutation:(BOOL)enable {
    [self.mTRTCCloud enableAudioVolumeEvaluation:enable ? 300 : 0];
}

#pragma mark - private method
- (void)internalEnterRoom{
    if (self.mTRTCParms) {
        self.mTRTCCloud.delegate = self;
        [self.mTRTCCloud enterRoom:self.mTRTCParms appScene:TRTCAppSceneVoiceChatRoom];
    }
}

- (BOOL)canDelegateResponseMethod:(SEL)method {
    return self.delegate && [self.delegate respondsToSelector:method];
}

#pragma mark - TRTCCloudDelegate
- (void)onEnterRoom:(NSInteger)result{
    TRTCLog(@"on enter trtc room. result:%ld", (long)result);
    if (result > 0) {
        self.isInRoom = YES;
        if (self.enterRoomCallback) {
            self.enterRoomCallback(0, @"enter trtc room success.");
        }
    } else {
        self.isInRoom = NO;
        if (self.enterRoomCallback) {
            self.enterRoomCallback((int)result, @"enter trtc room fail.");
        }
    }
    self.enterRoomCallback = nil;
}

- (void)onExitRoom:(NSInteger)reason {
    TRTCLog(@"on exit trtc room. reslut: %ld", (long)reason);
    self.isInRoom = NO;
    if (self.exitRoomCallback) {
        self.exitRoomCallback(0, @"exite room success");
    }
    self.exitRoomCallback = nil;
}

- (void)onRemoteUserEnterRoom:(NSString *)userId {
    TRTCLog(@"on user enter, userid: %@", userId);
    if ([self canDelegateResponseMethod:@selector(onTRTCAnchorEnter:)]) {
        [self.delegate onTRTCAnchorEnter:userId];
    }
}

- (void)onRemoteUserLeaveRoom:(NSString *)userId reason:(NSInteger)reason {
    if ([self canDelegateResponseMethod:@selector(onTRTCAnchorExit:)]) {
        [self.delegate onTRTCAnchorExit:userId];
    }
}

- (void)onUserAudioAvailable:(NSString *)userId available:(BOOL)available {
    if ([self canDelegateResponseMethod:@selector(onTRTCAudioAvailable:available:)]) {
        [self.delegate onTRTCAudioAvailable:userId available:available];
    }
}

- (void)onError:(TXLiteAVError)errCode errMsg:(NSString *)errMsg extInfo:(NSDictionary *)extInfo{
    if ([self canDelegateResponseMethod:@selector(onError:message:)]) {
        [self.delegate onError:errCode message:errMsg];
    }
}

- (void)onNetworkQuality:(TRTCQualityInfo *)localQuality remoteQuality:(NSArray<TRTCQualityInfo *> *)remoteQuality {
    if ([self canDelegateResponseMethod:@selector(onNetworkQuality:remoteQuality:)]) {
        [self.delegate onNetWorkQuality:localQuality arrayList:remoteQuality];
    }
}

- (void)onUserVoiceVolume:(NSArray<TRTCVolumeInfo *> *)userVolumes totalVolume:(NSInteger)totalVolume {
    if ([self canDelegateResponseMethod:@selector(onUserVoiceVolume:totalVolume:)]) {
        [self.delegate onUserVoiceVolume:userVolumes totalVolume:totalVolume];
    }
}

- (void)onSetMixTranscodingConfig:(int)err errMsg:(NSString *)errMsg{
    TRTCLog(@"on set mix transcoding, code:%d, msg: %@", err, errMsg);
}

@end

File: iOS/TRTCScenesDemo/TXLiteAVDemo/TRTCVoiceRoomDemo/model/Impl/trtc/VoiceRoomTRTCService.swift
//
//  VoiceRoomTRTCService.swift
//  TRTCVoiceRoomDemo
//
//  Created by abyyxwang on 2020/6/11.
//  Copyright © 2020 tencent. All rights reserved.
//

import Foundation

public protocol VoiceRoomTRTCServiceDelegate: class {
    func onTRTCAnchorEnter(userId: String)
    func onTRTCAnchorExit(userId: String)
    func onTRTCAudioAvailable(userId: String, available: Bool)
    func onError(errorCode: Int, message: String)
    func onNetWorkQuality(trtcQuality: TRTCQualityInfo, arrayList: [TRTCQualityInfo])
    func onUserVoiceVolume(userVolumes: [TRTCVolumeInfo], totalVolume: Int)
}

public class VoiceRoomTRTCService: NSObject {
    static let PLAY_TIME_OUT: Int = 5000
    
    private static let instance = VoiceRoomTRTCService.init()
    private override init() {}
    
    public static func getInstance() -> VoiceRoomTRTCService {
        return VoiceRoomTRTCService.instance
    }
    
    private var mTRTCCloud: TRTCCloud {
        return TRTCCloud.sharedInstance()
    }
    
    private var isInRoom: Bool = false
    private weak var delegate: VoiceRoomTRTCServiceDelegate?
    
    private var userId: String = ""
    private var roomId: String = ""
    private var mTRTCParms: TRTCParams?
    private var enterRoomCallback: TXCallback?
    private var exitRoomCallback: TXCallback?
    
    public func setDelegate(_ delegate: VoiceRoomTRTCServiceDelegate) {
        self.delegate = delegate
    }
    
    public func enterRoom(sdkAppId: Int32, roomId: String, userId: String, userSign: String, role: Int, callback: TXCallback?) {
        guard sdkAppId != 0 && roomId != "" && userId != "" && userSign != "" else {
            TRTCLog.out("error: enter trtc room fail. params invalid. room id:\(roomId), userId:\(userId), userSig is empty: \(userSign == "")")
            callback?(-1, "enter trtc room fail.")
            return
        }
        guard let roomIdIntValue = UInt32.init(roomId) else {
            TRTCLog.out("error: enter trtc room fail. params invalid. room id:\(roomId))")
            callback?(-1, "enter trtc room fail.")
            return
        }
        self.userId = userId
        self.roomId = roomId
        self.enterRoomCallback = callback
        TRTCLog.out("enter room. app id :\(sdkAppId). room id: \(roomId), userID: \(userId)")
        let trtcParams = TRTCParams.init()
        trtcParams.sdkAppId = UInt32(sdkAppId)
        trtcParams.userId = userId
        trtcParams.userSig = userSign
        trtcParams.role = TRTCRoleType.init(rawValue: role) ?? TRTCRoleType.audience
        trtcParams.roomId = roomIdIntValue
        mTRTCParms = trtcParams
        internalEnterRoom()
    }
    
    public func exitRoom(callback: TXCallback?) {
        TRTCLog.out("exit room.")
        userId = ""
        mTRTCParms = nil
        enterRoomCallback = nil
        exitRoomCallback = callback
        mTRTCCloud.exitRoom()
    }
    
    public func muteLocalAudio(isMute: Bool) {
        TRTCLog.out("mute local audio, mute: \(isMute)")
        mTRTCCloud.muteLocalAudio(isMute)
    }
    
    public func muteRemoteAudio(userId: String, mute: Bool) {
        TRTCLog.out("mute remote audio, user id: \(userId), mute: \(mute)")
        mTRTCCloud.muteRemoteAudio(userId, mute: mute)
    }
    
    public func muteAllRemoteAudio(isMute: Bool) {
        TRTCLog.out("mute all remote audio, mute:\(isMute)")
        mTRTCCloud.muteAllRemoteAudio(isMute)
    }
    
    public func setAuidoQuality(quality: Int) {
        mTRTCCloud.setAudioQuality(TRTCAudioQuality(rawValue: quality) ?? TRTCAudioQuality.default)
    }
    
    public func startMicrophone() {
        mTRTCCloud.startLocalAudio()
    }
    
    public func stopMicrophone() {
        mTRTCCloud.stopLocalAudio()
    }
    
    public func switchToAnchor() {
        mTRTCCloud.switch(TRTCRoleType.anchor)
        mTRTCCloud.startLocalAudio()
    }
    
    public func switchToAudience() {
        mTRTCCloud.stopLocalAudio()
        mTRTCCloud.switch(TRTCRoleType.audience)
    }
    
    public func setSpeaker(useSpeaker: Bool) {
        mTRTCCloud.setAudioRoute(useSpeaker ? .modeSpeakerphone : .modeEarpiece)
    }
    
    public func setAudioCaptureVolume(volume: Int) {
        mTRTCCloud.setAudioCaptureVolume(volume)
    }
    
    public func setAudioPlayoutVolume(volume: Int) {
        mTRTCCloud.setAudioPlayoutVolume(volume)
    }
    
    public func startFileDumping(trtcAudioRecordingParams: TRTCAudioRecordingParams) {
        mTRTCCloud.startAudioRecording(trtcAudioRecordingParams)
    }
    
    public func stopFileDumping() {
        mTRTCCloud.stopAudioRecording()
    }
    
    public func enableAudioEvalutaion(enable: Bool) {
        mTRTCCloud.enableAudioVolumeEvaluation(enable ? 300 : 0)
    }
    
    
}

extension VoiceRoomTRTCService {
    private func internalEnterRoom() {
        guard let parms = mTRTCParms else { return }
        mTRTCCloud.delegate = self
        mTRTCCloud.enterRoom(parms, appScene: TRTCAppScene.voiceChatRoom)
    }
}

extension VoiceRoomTRTCService: TRTCCloudDelegate {
    public func onEnterRoom(_ result: Int) {
        TRTCLog.out("on enter room. result: \(result)")
        if result > 0 {
            isInRoom = true
            enterRoomCallback?(0, "enter room success")
        } else {
            isInRoom = false
            enterRoomCallback?(Int32(result), "enter room fail")
        }
        enterRoomCallback = nil
    }
    
    public func onExitRoom(_ reason: Int) {
        TRTCLog.out("on exit room. result: \(reason)")
        isInRoom = false
        exitRoomCallback?(0, "exit room success")
        exitRoomCallback = nil
    }
    
    public func onRemoteUserEnterRoom(_ userId: String) {
        TRTCLog.out("on user enter, user id: \(userId)")
        delegate?.onTRTCAnchorEnter(userId: userId)
    }
    
    public func onRemoteUserLeaveRoom(_ userId: String, reason: Int) {
        TRTCLog.out("on user exit. id: \(userId)")
        delegate?.onTRTCAnchorExit(userId: userId)
    }
    
    public func onUserAudioAvailable(_ userId: String, available: Bool) {
        TRTCLog.out("on user audio available, usr id: \(userId), available:\(available)")
        delegate?.onTRTCAudioAvailable(userId: userId, available: available)
    }
    
    public func onError(_ errCode: TXLiteAVError, errMsg: String?, extInfo: [AnyHashable : Any]?) {
        TRTCLog.out("on error: \(String(describing: errMsg)), code: \(errCode)")
        delegate?.onError(errorCode: Int(errCode.rawValue), message: errMsg ?? "")
    }
    
    public func onNetworkQuality(_ localQuality: TRTCQualityInfo, remoteQuality: [TRTCQualityInfo]) {
        delegate?.onNetWorkQuality(trtcQuality: localQuality, arrayList: remoteQuality)
    }
    
    public func onUserVoiceVolume(_ userVolumes: [TRTCVolumeInfo], totalVolume: Int) {
        delegate?.onUserVoiceVolume(userVolumes: userVolumes, totalVolume: totalVolume)
    }
    
    public func onSetMixTranscodingConfig(_ err: Int32, errMsg: String) {
        TRTCLog.out("on set mix transcoding, code: \(err), msg:\(errMsg)")
    }
    
}
